### PR TITLE
Add multiple transactions to onchain analysis and fix suicide module

### LINKED ISSUE #665: Incorrect behavior in on-chain analysis
## Description

Mythril behaves incorrectly when analyzing an on-chain instance of [Ownable.sol](https://github.com/ConsenSys/devcon4-playground/blob/master/exercise4/contracts/Ownable.sol) as opposed to the source code.

## How to Reproduce

Analyzing the source produces the correct transaction sequence (call `Ownab1e()`, the `kill()`):

```
$ myth -msuicide --max-transaction-count 2 --verbose-report -x Ownable.sol
==== Unchecked SUICIDE ====
SWC ID: 106
Type: Warning
Contract: Ownable
Function name: kill()
PC address: 308
A reachable SUICIDE instruction was detected. The remaining Ether is sent to the caller's address.

--------------------
In file: Ownable.sol:13

selfdestruct(msg.sender)

--------------------
--------------------
DEBUGGING INFORMATION:

Transaction Sequence: {'2': {'calldata': '0x2f111e8f', 'call_value': '0x0', 'caller': '0x0000000000000000000000000000000000000000000000000000000000000002'}, '3': {'calldata': '0x41c0e1b51010101010', 'call_value': '0x0', 'caller': '0x0000000000000000000000000000000000000000000000000000000000000002'}}
```
Running against an instance deployed on Ropsten seems to miss the constraint on the contract owner:

```
$ myth --rpc infura-ropsten -msuicide --max-transaction-count 2 --verbose-report -xa 0x4d1734ec536e6dafa3d4ceee532d5b5c8d45fe22
==== Unchecked SUICIDE ====
SWC ID: 106
Type: Warning
Contract: 0x4d1734ec536e6dafa3d4ceee532d5b5c8d45fe22
Function name: kill()
PC address: 185
A reachable SUICIDE instruction was detected. The remaining Ether is sent to the caller's address.

--------------------
--------------------
DEBUGGING INFORMATION:

Transaction Sequence: {'1': {'calldata': '0x41c0e1b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5', 'call_value': '0x0', 'caller': '0x0000000000000000000000000000000000000000000000000000000000000000'}}
```

## Expected behavior

I would expect Mythril to automatically the constraint on the suicide instruction, and also determine the current value of owner from storage. It should then correctly determine that there exists a tx that changes the owner and a follow-up tx that kills the contract.

## Environment

- Mythril version: v0.19.3
- Solidity compiler and version: 0.4.24+commit.e67f0147.Darwin.appleclang
- Python version:  3.6.3
- OS and Version: Mac OS Mojave

### PR DESCRIPTION
**Pull request by norhh**
fixes #665 .

- Add mult transaction support to onchain analysis
- Take care of constraints for onchain analysis in suicide module

### CODE REVIEW

File: mythril/analysis/modules/suicide.py [hunk 1]
```diff
@@ -54,15 +55,25 @@ def _analyze_state(state, node):
...
-    if len(state.world_state.transaction_sequence) > 1:
-        creator = state.world_state.transaction_sequence[0].caller
-        for transaction in state.world_state.transaction_sequence[1:]:
-            not_creator_constraints.append(
-                Not(Extract(159, 0, transaction.caller) == Extract(159, 0, creator))
-            )
-            not_creator_constraints.append(
-                Not(Extract(159, 0, transaction.caller) == 0)
-            )
+    if len(state.world_state.transaction_sequence) >= 1:
```
Comment (JoranHonig) [inline mythril/analysis/modules/suicide.py]: this simplifies to True :sweat_smile:

File: mythril/analysis/modules/suicide.py [hunk 2]
```diff
@@ -54,15 +55,25 @@ def _analyze_state(state, node):
...
+        creator = None
```
Comment (JoranHonig) [inline mythril/analysis/modules/suicide.py]: This statement isn't necessary right?

File: mythril/analysis/modules/suicide.py [hunk 3]
```diff
@@ -54,15 +55,25 @@ def _analyze_state(state, node):
...
+                not_creator_constraints.append(
+                    Not(Extract(159, 0, transaction.caller) == Extract(159, 0, creator))
+                )
+                not_creator_constraints.append(
+                    Not(Extract(159, 0, transaction.caller) == 0)
+                )
+        else:
+            for transaction in state.world_state.transaction_sequence:
+                not_creator_constraints.append(
+                    Not(Extract(159, 0, transaction.caller) == 0)
```
Comment (JoranHonig) [inline mythril/analysis/modules/suicide.py]: The problem here is that the owner can suicide the contract, and the module won't know

Comment (JoranHonig) [reply mythril/analysis/modules/suicide.py]: I think for on chain analysis the specification should be different, namely something like:
If somebody with a specific address can trigger a sequence of transactions that triggers the suicide then the contract is vulnerable. (We can even make this configurable)

Comment (JoranHonig) [reply mythril/analysis/modules/suicide.py]: Or what previously happened, check if there is a constraint on the caller and check if you can change that constraint

File: mythril/analysis/modules/suicide.py (hunk 2 above)
Comment (norhh) [reply mythril/analysis/modules/suicide.py]: without this `creator` will be uninitialized if the `if` condition fails right?
I think this should have been in it's corresponding `else`, but for some reason i didn't want extra if-else there, should i add it in `else`?

File: mythril/analysis/modules/suicide.py (hunk 1 above)
Comment (norhh) [reply mythril/analysis/modules/suicide.py]: I wrote it for safety as I was directly indexing the transaction sequence, but looks like an assert would be better there.

File: mythril/analysis/modules/suicide.py (hunk 3 above)
Comment (norhh) [reply mythril/analysis/modules/suicide.py]: previous implementation had a vague check on changeability for caller, should we do something regarding that at this moment?

**Comment by b-mueller** [comment]
Removing the "high priority" label since even though I'd like to have this ASAP, the on-chain analysis is more of a pet project ;)

**Comment by b-mueller** [comment]
I'm getting some interesting results but it isn't completely working.

**Form source with --max-tx-count=1 (correct result):**

```
$ $MYTH -msuicide -x ~/Desktop/Ownable.sol 
The analysis was completed successfully. No issues were detected.

```

**From source with --max-tx-count=2 (correct result):**

```
$ $MYTH -msuicide --max-transaction-count 2 -x ~/Desktop/Ownable.sol --verbose-report
==== Unchecked SUICIDE ====
SWC ID: 106
Type: Warning
Contract: Ownable
Function name: kill()
PC address: 406
A reachable SUICIDE instruction was detected. The remaining Ether is sent to the caller's address.

--------------------
In file: /Users/bernhardmueller/Desktop/Ownable.sol:13

selfdestruct(msg.sender)

--------------------
--------------------
DEBUGGING INFORMATION:

Transaction Sequence: {'3': {'caller': '0x0000000000000000000000000000000000000000000000010000000000000000', 'call_value': '0x0', 'calldata': '0x41c0e1b5b5b5b5b5b5'}, '2': {'caller': '0x0000000000000000000000000000000000000000000000010000000000000000', 'call_value': '0x0', 'calldata': '0x2f111e8f8f8f8f8(...)'}}
```

**On Ropsten with --max-tx-count=1 (correct result):**

```
$MYTH -msuicide --rpc infura-ropsten -xa 0x345ca3e014aaf5dca488057592ee47305d9b3e10
The analysis was completed successfully. No issues were detected.
```

**On Ropsten with --max-tx-count=2 (times out, or at least doesn't return a result after 5 minutes):**

```
$ $MYTH -msuicide --rpc infura-ropsten --max-transaction-count 2 -xa 0x345ca3e014aaf5dca488057592ee47305d9b3e10 -v1
root[2569] INFO Using RPC settings: ('ropsten.infura.io', 443, True)
root[2569] INFO LASER EVM initialized with dynamic loader: None
root[2569] INFO Starting message call transaction to 0x345ca3e014aaf5dca488057592ee47305d9b3e10
root[2569] INFO Starting message call transaction, iteration: 0
root[2569] INFO Starting message call transaction, iteration: 1
```

**On local Ganache (returns the same false positive as before):**

```
$ $MYTH --rpc localhost:7545 -xa 0x345ca3e014aaf5dca488057592ee47305d9b3e10 --verbose-report
==== Unchecked SUICIDE ====
SWC ID: 106
Type: Warning
Contract: 0x345ca3e014aaf5dca488057592ee47305d9b3e10
Function name: kill()
PC address: 406
A reachable SUICIDE instruction was detected. The remaining Ether is sent to the caller's address.

--------------------
--------------------
DEBUGGING INFORMATION:

Transaction Sequence: {'1': {'calldata': '0x41c0e1b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5b5', 'call_value': '0x0', 'caller': '0x0000000000000000000000000000000000000000000000000000000000000001'}}

 $MYTH --rpc localhost:7545 -a 0x345ca3e014aaf5dca488057592ee47305d9b3e10 --storage 0
0: 0x00

```

I wonder why there is such a large difference in performance between analysis from source and on-chain code, it's the same byte code after all. Is it because many RPC requests need to be sent?

**Comment by norhh** [comment]
it's still pending https://github.com/ConsenSys/mythril-classic/pull/646

**Comment by b-mueller** [comment]
With the -l flag, I get the correct results from Ganache. Still testing on Ropsten.

**Comment by b-mueller** [comment]
Ok. With -l added the analysis runs much faster than without, but the issue is not detected:

```
$ $MYTH -msuicide --rpc infura-ropsten -xla 0x345ca3e014aaf5dca488057592ee47305d9b3e10 --max-transaction-count 3 -v1
root[4467] INFO Using RPC settings: ('ropsten.infura.io', 443, True)
root[4467] INFO LASER EVM initialized with dynamic loader: <mythril.support.loader.DynLoader object at 0x10676ab00>
root[4467] INFO Starting message call transaction to 0x345ca3e014aaf5dca488057592ee47305d9b3e10
root[4467] INFO Starting message call transaction, iteration: 0
root[4467] INFO Starting message call transaction, iteration: 1
root[4467] INFO Finished symbolic execution
root[4467] INFO 327 nodes, 326 edges, 5419 total states
root[4467] INFO Achieved 98.70% coverage for code: 0x60606040526004361061006d576000357c0100000000000000000000000000000000000000000000000000000000900463ffffffff1680630429e372146100725780635eff7dae146101195780638eb5234c14610183578063ca8ab933146101e6578063d69e1e2d1461020f575b600080fd5b341561007d57600080fd5b6100a9600480803573ffffffffffffffffffffffffffffffffffffffff16906020019091905050610289565b60405180848152602001836fffffffffffffffffffffffffffffffff19166fffffffffffffffffffffffffffffffff19168152602001826fffffffffffffffffffffffffffffffff19166fffffffffffffffffffffffffffffffff19168152602001935050505060405180910390f35b341561012457600080fd5b61012c61039e565b6040518080602001828103825283818151815260200191508051906020019060200280838360005b8381101561016f578082015181840152602081019050610154565b505050509050019250505060405180910390f35b341561018e57600080fd5b6101a46004808035906020019091905050610432565b604051808273ffffffffffffffffffffffffffffffffffffffff1673ffffffffffffffffffffffffffffffffffffffff16815260200191505060405180910390f35b34156101f157600080fd5b6101f9610471565b6040518082815260200191505060405180910390f35b341561021a57600080fd5b610287600480803573ffffffffffffffffffffffffffffffffffffffff169060200190919080359060200190919080356fffffffffffffffffffffffffffffffff19169060200190919080356fffffffffffffffffffffffffffffffff191690602001909190505061047e565b005b6000806000600160008573ffffffffffffffffffffffffffffffffffffffff1673ffffffffffffffffffffffffffffffffffffffff16815260200190815260200160002060000154600160008673ffffffffffffffffffffffffffffffffffffffff1673ffffffffffffffffffffffffffffffffffffffff16815260200190815260200160002060010160009054906101000a900470010000000000000000000000000000000002600160008773ffffffffffffffffffffffffffffffffffffffff1673ffffffffffffffffffffffffffffffffffffffff16815260200190815260200160002060010160109054906101000a9004700100000000000000000000000000000000029250925092509193909250565b6103a661069f565b600280548060200260200160405190810160405280929190818152602001828054801561042857602002820191906000526020600020905b8160009054906101000a900473ffffffffffffffffffffffffffffffffffffffff1673ffffffffffffffffffffffffffffffffffffffff16815260200190600101908083116103de575b5050505050905090565b60028181548110151561044157fe5b90600052602060002090016000915054906101000a900473ffffffffffffffffffffffffffffffffffffffff1681565b6000600280549050905090565b60008060009054906101000a900473ffffffffffffffffffffffffffffffffffffffff1673ffffffffffffffffffffffffffffffffffffffff163373ffffffffffffffffffffffffffffffffffffffff161415156104db57600080fd5b600160008673ffffffffffffffffffffffffffffffffffffffff1673ffffffffffffffffffffffffffffffffffffffff1681526020019081526020016000209050838160000181905550828160010160006101000a8154816fffffffffffffffffffffffffffffffff021916908370010000000000000000000000000000000090040217905550818160010160106101000a8154816fffffffffffffffffffffffffffffffff0219169083700100000000000000000000000000000000900402179055506001600280548060010182816105b591906106b3565b9160005260206000209001600088909190916101000a81548173ffffffffffffffffffffffffffffffffffffffff021916908373ffffffffffffffffffffffffffffffffffffffff16021790555050507fa60691fff33251eeea9783cf4ce933e6cd6710379c9532c1b3b061d023cb50b783838660405180846fffffffffffffffffffffffffffffffff19166fffffffffffffffffffffffffffffffff19168152602001836fffffffffffffffffffffffffffffffff19166fffffffffffffffffffffffffffffffff19168152602001828152602001935050505060405180910390a15050505050565b602060405190810160405280600081525090565b8154818355818115116106da578183600052602060002091820191016106d991906106df565b5b505050565b61070191905b808211156106fd5760008160009055506001016106e5565b5090565b905600a165627a7a723058201a3c37239f809cd91959951ea86dbcc624a8bbb739d012e2afff268c47f959e90029
root[4467] INFO Starting analysis
root[4467] INFO Executing <module 'suicide' from '/Users/bernhardmueller/Projects/mythril-classic/mythril/analysis/modules/suicide.py'>
The analysis was completed successfully. No issues were detected.

```

**Comment by norhh** [comment]
Can you mention the contract code, the contract's bytecode seems to be much bigger than the bytecode for Ownable.sol?

**Comment by b-mueller** [comment]
Hmm I think I mixed something up. Just deployed a new instance and it works :)

**Pull request merged**